Support hexadecimal literals as integers

wolffcm opened the pull request:
This changeset supports using hexadecimal literals (`X'00ABCD'`) to express integers in SQL.

This required some changes to HSQL to convert VARBINARY literals into BIGINTs based on context.  This is done for arithmetic (+, -, *, /) and logical (<. >, =, etc) expressions, and the new bitwise functions.

I decided *not* to support hexadecimal literals for the other, non-bitwise functions, since it seems really unlikely that a user would ever type something like `SQRT(X'0100')`.  The way HSQL handles the non-VoltDB specific functions this would require changes for each function.  I can file a ticket to support these other functions, if we feel there's a need for it.

paulmartel commented on src/frontend/org/voltdb/ParameterConverter.java:
```diff
@@ -108,8 +108,26 @@ else if (value != null) {
      * Given a string, covert it to a primitive type or return null.
+     *
+     * If the string value is a VARBINARY constant of the form X'00ABCD', the
+     * expected class is one of byte, short, int or long, then we interpret the
```
Add "and" for  "... and the expected class is..."

paulmartel commented on src/frontend/org/voltdb/ParameterConverter.java:
```diff
@@ -138,7 +156,24 @@ private static Object convertStringToPrimitive(String value, final Class<?> expe
+        //
+        // Do this only after trying to parse a decimal literal, which is the
+        // most common case.
+        if (isIntegralClass(expectedClz)) {
```
Not sure how much we care, but this code comes after the CSV-friendly code:

         // remove commas and escape chars
         value = thousandSeparator.matcher(value).replaceAll("");

We COULD preserve and use the original raw value for the x-literal case.
Do we want to allow/ignore or disallow commas and/or back-slashes in and around our x literals?

paulmartel replied:
Move this block into the catch block above to avoid for non-integer non-double.
In that context, isIntegralClass() could be simplified to a test for the only NumberFormatException that we don't retry as a possible x-literal: (expectedClz != double.class).

wolffcm replied:
I'm looking at the `thousandSeparator` code.  It doesn't seem to do anything with backslashes, it just removes commas.  

Removing commas arbitrarily can result in some strange behavior: for example, we'll parse `5.2,6e,-,32,,,` as a valid double value.  We don't seem to document this behavior in CSVLoader.  Might it be documented somewhere else?  I'm more inclined (as a separate ticket) to rip out the comma-replacement altogether so that we can just write in our docs, "CSVLoader parses numeric values using Java's `parseLong`, `parseDouble`, etc methods."

Of course, it's possible that someone is relying on this behavior now, so maybe the right thing would to do is to preserve the original value for when we try to parse a hex literal, so as to stanch this strange behavior.  So I guess that's what I'll do.

paulmartel commented on src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionLogical.java:
```diff
@@ -1699,4 +1703,29 @@ boolean reorderComparison(Session session) {
+     * If one child is an integer, and the other is a VARBINARY literal, try to convert the
+     * literal to an integer.
+     */
+    private boolean convertBinaryIntegerLiteral(Session session, Expression lhs, Expression rhs) {
```
With the hsql232 upgrade, I've been moving towards smurf-naming the attributes and methods we add(ed) with a volt prefix. It helps. So, voltConvertBinaryIntegerLiteral, voltMutateToBigintType, etc.

paulmartel commented on src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionLogical.java:
```diff
@@ -1699,4 +1703,29 @@ boolean reorderComparison(Session session) {
+        }
+
+        if (nonIntegralExpr.opType == OpTypes.VALUE && nonIntegralExpr.dataType.isBinaryType()) {
+            int whichChild = (nonIntegralExpr == rhs) ? RIGHT : LEFT;
```
set whichChild in the if/else block above, rather than reset here.

Should the "opType == OpTypes.VALUE && dataType.isBinaryType() and cast to ExpressionValue" preamble common to all of mutateToBigintType's callers be pushed down into it? That WOULD mean that mutateToBigintType would not be an instance method or maybe make it private with a static wrapper method that executes the preamble.

paulmartel commented on src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionValue.java:
```diff
@@ -97,4 +102,47 @@ protected String describe(Session session, int blanks) {
+     *
+     * These are the VoltDB classes that handle hex literal constants:
+     *   voltdb.ParameterConverter
+     *   voltdb.expressions.ConstantValueExpression
```
It's not clear from your comment what the relationship is here -- that each of these code paths should consistently implement the same cases. sqlcmd and VMC's exec param parsing should also be mentioned (though AFAIK the latter's x-literal support is TBD).

paulmartel commented on src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionValue.java:
```diff
@@ -97,4 +102,47 @@ protected String describe(Session session, int blanks) {
+
+        byte[] data = ((BinaryData)valueData).getBytes();
+        if (data == null || data.length <= 0 || data.length > 16) {
+            return false;
```
It seems a little sad and strange that all of these different failure cases are summed up with a false return value -- and even that return value is not checked by one of your callers. That suggests that the resulting error messages are likely to be pretty generic. If at any point we know that we are in an unrecoverable situation (vs. some kind of a trial interpretation) we'd be better off throwing more specific errors.

wolffcm replied:
My thinking here was that if we can't convert the literal to an integer (too many or too few digits), then we fall back to treating it as a varbinary literal, so the error message would be something like "incompatible data types in combination."

I think you're right that this would be confusing, so I can instead return the error "malformed numeric constant" instead, which is a little better.

paulmartel commented on tests/frontend/org/voltdb/compiler/TestVoltCompilerErrorMsgs.java:
```diff
@@ -209,29 +215,42 @@ public void testErrorOnLimitPartitionRows() throws Exception {
+
+        ddlErrorTest("invalid format for a constant decimal value",
+                "create procedure insHex as insert into blah (decval) values (x'80');");
+
```
Test that it is possible (albeit inconvenient) to set the high bit of a TINYINT by with 14 Fs + an '8' or higher + one more digit.

paulmartel replied:
I see that you have something like this covered in TestHexLiteralsSuite.

paulmartel commented on tests/frontend/org/voltdb/compiler/TestVoltCompilerErrorMsgs.java:
```diff
@@ -39,7 +39,10 @@ private void statementTest(String feature, boolean expectError, boolean testStmt
+            "sval varchar(255) not null," +
+            "tinyval tinyint," +
+            "floatval float," +
+            "decval decimal" +
```
Add a table that demonstrates valid usage of x-literal default values (is this supported? -- does it work compatibly in hsql backend?).
Add a test demonstrating ddl error messages resulting from invalid hex literal default value cases.

paulmartel commented:
End of review round.

wolffcm commented:
@paulmartel I added another commit that addresses your comments, and supports default values.  It turns out that there is explicit code in HSQL for unary minus in default values, so I didn't need to do anything special to support `DEFAULT -X'7F'`.
